README
Authen::CAS::Client is an implementation of a client for use with
JA-SIG's Central Authentication Service.


More information about the CAS protocol can be found
at the JA-SIG CAS homepage:

http://www.ja-sig.org/products/cas/


To install this module, use the following steps:

  perl Makefile.PL
  make
  make test
  make install


CHANGES
changes from 0.03 to 0.04
-------------------------

  * changed fatal behavior to throw strings instead of objects


changes from 0.02 to 0.03
-------------------------

  * added more tests that were previously not included

  * changed to Module::Install for dist packaging


changes from 0.01 to 0.02
-------------------------

  * added Perl 5.6.1 dependency

  * changed Authen::CAS::Client::Response is_* methods to be more consistent
    across all response types

  * added tests for better code coverage

  * various minor internal fixups to maintain consistency


t/MockUserAgent.pm
package Test::MockUserAgent;

use strict;
use warnings;

use base qw/ Test::MockObject /;

use HTTP::Response;

BEGIN { Test::MockObject->fake_module( 'LWP::UserAgent' ); }

sub new {
  my ( $class ) = @_;

  my $self = $class->SUPER::new();
  $self->fake_new( 'LWP::UserAgent' );
  $self->mock( request => sub { my ( $self ) = @_; $self->{_response} } );

  return $self;
}

sub _response {
  my ( $self, $code, $content ) = @_;

  my $response = HTTP::Response->new( $code );
  $response->content( $content );
  $response->message( 'Fake HTTP Response' );

  $self->{_response} = $response;
}

1;


lib/Authen/CAS/Client/Response.pm
require 5.006_001;

use strict;
use warnings;

# $Id: Response.pm 32 2009-05-05 19:15:42Z jhord $

#======================================================================
# Authen::CAS::Client::Response
#
package Authen::CAS::Client::Response;

our $VERSION = '0.02';

sub new {
  my ( $class, %args ) = @_;

  my $self = { };
  $self->{$_} = $args{$_}
    for keys %args;

  bless $self, $class
}

sub is_error   { my ( $self ) = @_; ! defined $self->{_ok} }
sub is_failure { my ( $self ) = @_;   defined $self->{_ok} && ! $self->{_ok} }
sub is_success { my ( $self ) = @_;   defined $self->{_ok} &&   $self->{_ok} }


#======================================================================
# Authen::CAS::Client::Response::Error
#
package Authen::CAS::Client::Response::Error;

use base qw/ Authen::CAS::Client::Response /;

sub new {
  my ( $class, %args ) = @_;

  $class->SUPER::new(
    _ok   => undef,
    error => defined $args{error} ? $args{error} : 'An internal error ocurred',
  );
}

sub error { my ( $self ) = @_; $self->{error} }


#======================================================================
# Authen::CAS::Client::Response::Failure
#
package Authen::CAS::Client::Response::Failure;

use base qw/ Authen::CAS::Client::Response /;

sub new {
  my ( $class, %args ) = @_;

  $class->SUPER::new(
    _ok      => 0,
    code     => $args{code},
    message  => defined $args{message} ? $args{message} : '',
  );
}

sub code    { my ( $self ) = @_; $self->{code} }
sub message { my ( $self ) = @_; $self->{message} }


#======================================================================
# Authen::CAS::Client::Response::AuthFailure
#
package Authen::CAS::Client::Response::AuthFailure;

use base qw/ Authen::CAS::Client::Response::Failure /;


#======================================================================
# Authen::CAS::Client::Response::ProxyFailure
#
package Authen::CAS::Client::Response::ProxyFailure;

use base qw/ Authen::CAS::Client::Response::Failure /;


#======================================================================
# Authen::CAS::Client::Response::Success
#
package Authen::CAS::Client::Response::Success;

use base qw/ Authen::CAS::Client::Response /;

sub _ATTR () { }

sub new {
  my ( $class, %args ) = @_;

  my %attr = $class->_ATTR;
  for ( keys %attr ) {
    $attr{$_} = $args{$_}
      if exists $args{$_};
  }

  $class->SUPER::new( _ok => 1, %attr );
}


#======================================================================
# Authen::CAS::Client::Response::AuthSuccess
#
package Authen::CAS::Client::Response::AuthSuccess;

use base qw/ Authen::CAS::Client::Response::Success /;

sub _ATTR () { ( user => undef, iou => undef, proxies => [ ] ) }

sub user    { my ( $self ) = @_; $self->{user} }
sub iou     { my ( $self ) = @_; $self->{iou} }
sub proxies { my ( $self ) = @_; wantarray ? @{ $self->{proxies} } : [ @{ $self->{proxies} } ] }


#======================================================================
# Authen::CAS::Client::Response::ProxySuccess
#
package Authen::CAS::Client::Response::ProxySuccess;

use base qw/ Authen::CAS::Client::Response::Success /;

sub _ATTR() { ( pt => undef ) }

sub proxy_ticket { my ( $self ) = @_; $self->{pt} }


1;

__END__

=head1 NAME

Authen::CAS::Client::Response - A set of classes for implementing
responses from a CAS server

=head1 DESCRIPTION

C<Authen::CAS::Client::Response> implements a base class that is used to
build a hierarchy of response objects that are returned from methods in
L<Authen::CAS::Client>.  Most response objects are meant to encapsulate
a type of response from a CAS server.

=head1 CLASSES AND METHODS

=head2 Authen::CAS::Client::Response

C<Authen::CAS::Client::Response> is the base class from which all other
response classes inherit.  As such it is very primitive and is never
used directly.

=over 2

=item new %ARGS

new() creates an instance of an C<Authen::CAS::Client::Response> object
and assigns its data members according to the values in C<%ARGS>.

=item is_error

is_error() returns true if the response represents an error object.

=item is_failure

is_failure() returns true if the response represents a failure object.

=item is_success

is_success() returns true if the response represents a success object.

=back

=head2 Authen::CAS::Client::Response::Error

C<Authen::CAS::Client::Response::Error> is used when an error occurs that
prevents further processing of a request.  This would include not being able
connect to the CAS server, receiving an unexpected response from the server
or being unable to correctly parse the server's response according to the
guidelines in the CAS protocol specification.

=over 2

=item new $ERROR

new() creates an instance of an C<Authen::CAS::Client::Response::Error>
object.  C<$ERROR> is the error string.

=item error

error() returns the error string.

=back

=head2 Authen::CAS::Client::Response::Failure

C<Authen::CAS::Client::Response::Failure> is used as a base class for other
failure responses.  These correspond to the C<cas:authenticationFailure> and
C<cas:proxyFailure> server responses outlined in the CAS protocol
specification.

=over 2

=item new $CODE, $MESSAGE

new() creates an instance of an C<Authen::CAS::Client::Response::Failure>
object.  C<$CODE> is the failure code.  C<$MESSAGE> is the failure message.

=item code

code() returns the failure code.

=item message

message() returns the failure message.

=back

=head2 Authen::CAS::Client::Response::AuthFailure

C<Authen::CAS::Client::Response::AuthFailure> is used when a
C<cas:authenticationFailure> response is received from the CAS server
during a validation attempt.  When using the CAS v2.0 protocol,
C<$CODE> and C<$MESSAGE> are set according to what is parsed from the
server response.  When using the CAS v1.0 protocol, C<$CODE> is set
to 'V10_AUTH_FAILURE' and C<$MESSAGE> is set to the empty string.

No additional methods are defined.

=head2 Authen::CAS::Client::Response::ProxyFailure

C<Authen::CAS::Client::Response::ProxyFailure> is used when a
<cas:proxyFailure> response is received from the CAS server
during a proxy attempt.  C<$CODE> and C<$MESSAGE> are set according
to what is parsed from the server response.

=head2 Authen::CAS::Client::Response::Success

C<Authen::CAS::Client::Response::Success> is used as base class for other
success responses.  These correspond to the C<cas:authenticationSuccess> and
C<cas:proxySuccess> server responses.

No additional methods are defined.

=head2 Authen::CAS::Client::Response::AuthSuccess

C<Authen::CAS::Client::Response::AuthSuccess> is used when a
C<cas:authenticationSuccess> response is received from the CAS server
during a validation attempt.

=over 2

=item user

user() returns the user name that was contained in the server response.

=item iou

iou() returns the proxy granting ticket IOU, if it was present in the
server response.  Otherwise it is set to C<undef>.

=item proxies

proxies() returns the list of proxies present in the server response.  If
no proxies are found, an empty list is returned.  In scalar context an
array reference will be returned instead.

=back

=head2 Authen::CAS::Client::Response::ProxySuccess

C<Authen::CAS::Client::Response::ProxySuccess> is used when a
C<cas:proxySuccess> response is received from the CAS server during
a proxy attempt.

=over 2

=item proxy_ticket

proxy_ticket() returns the proxy ticket that was contained in the
server response.

=back

=head1 BUGS

None are known at this time, but if you find one, please feel free to
submit a report to the author.

=head1 AUTHOR

jason hord E<lt>pravus@cpan.orgE<gt>

=head1 SEE ALSO

=over 2

=item L<Authen::CAS::Client>

=back

=head1 COPYRIGHT

Copyright (c) 2007-2009, jason hord

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:

=over 2

=item *

Redistributions of source code must retain the above copyright
notice, this list of conditions and the following disclaimer.

=item *

Redistributions in binary form must reproduce the above
copyright notice, this list of conditions and the following
disclaimer in the documentation and/or other materials provided
with the distribution.

=back

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut


lib/Authen/CAS/Client.pm
package Authen::CAS::Client;

# $Id: Client.pm 32 2009-05-05 19:15:42Z jhord $

require 5.006_001;

use strict;
use warnings;

use LWP::UserAgent;
use URI;
use URI::QueryParam;
use XML::LibXML;
use Authen::CAS::Client::Response;

our $VERSION = '0.04';

#======================================================================
# constructor
#

sub new {
  my ( $class, $cas, %args ) = @_;

  my $self = {
    _cas   => URI->new( $cas ),
    _ua    => LWP::UserAgent->new( agent => "WWW-CAS-Service/$VERSION" ),
    _fatal => $args{fatal} ? 1 : 0,
  };

  bless $self, $class;
}


#======================================================================
# private methods
#

sub _error {
  my ( $self, $error ) = @_;

  die $error
    if $self->{_fatal};

  Authen::CAS::Client::Response::Error->new( error => $error );
}

sub _parse_auth_response {
  my ( $self, $xml ) = @_;

  my $root = XML::LibXML->new()->parse_string( $xml );

  my ( $node, $response );

  if( $node = $root->find( '/cas:serviceResponse/cas:authenticationSuccess' )->get_node( 1 ) ) {
    $response = eval {
      my $user = $node->find( './cas:user' )->get_node( 1 )->textContent();

      my $iou = $node->find( './cas:proxyGrantingTicket' )->get_node( 1 );
      $iou = $iou->textContent()
        if( defined $iou );

      my $proxies = $node->findnodes( './cas:proxies/cas:proxy' );
      $proxies = [ map $_->textContent(), @$proxies ]
        if defined @$proxies;

      Authen::CAS::Client::Response::AuthSuccess->new(
        user    => $user,
        iou     => $iou,
        proxies => $proxies,
      );
    };
    $response = $self->_error( 'Failed to parse authentication success response' )
      if $@;
  }
  elsif( $node = $root->find( '/cas:serviceResponse/cas:authenticationFailure' )->get_node( 1 ) ) {
    $response = eval {
      die
        unless $node->hasAttribute( 'code' );
      my $code = $node->getAttribute( 'code' );
      
      my $message = $node->textContent();
      if( defined $message ) {
        s/^\s+//, s/\s+\z//
          for $message;
      }

      Authen::CAS::Client::Response::AuthFailure->new(
        code    => $code,
        message => $message,
      );
    };
    $response = $self->_error( 'Failed to parse authentication failure response' )
      if $@;
  }
  else {
    die;
  }

  return $response;
}

sub _parse_proxy_response {
  my ( $self, $xml ) = @_;

  my $root = XML::LibXML->new()->parse_string( $xml );

  my ( $node, $response );

  if( $node = $root->find( '/cas:serviceResponse/cas:proxySuccess' )->get_node( 1 ) ) {
    $response = eval {
      my $pt = $node->find( './cas:proxyTicket' )->get_node( 1 )->textContent();

      Authen::CAS::Client::Response::ProxySuccess->new(
        pt => $pt,
      );
    };
    $response = $self->_error( 'Failed to parse proxy success response' )
      if $@;
  }
  elsif( $node = $root->find( '/cas:serviceResponse/cas:proxyFailure' )->get_node( 1 ) ) {
    $response = eval {
      die
        unless $node->hasAttribute( 'code' );
      my $code = $node->getAttribute( 'code' );
      
      my $message = $node->textContent();
      if( defined $message ) {
        s/^\s+//, s/\s+\z//
          for $message;
      }

      Authen::CAS::Client::Response::ProxyFailure->new(
        code    => $code,
        message => $message,
      );
    };
    $response = $self->_error( 'Failed to parse proxy failure response' )
      if $@;
  }
  else {
    die;
  }

  return $response;
}

sub _server_request {
  my ( $self, $path, $params ) = @_;

  my $url      = $self->_url( $path, $params )->canonical();
  my $response = $self->{_ua}->get( $url );

  unless( $response->is_success() ) {
    return $self->_error(
      'HTTP request failed: ' . $response->code() . ': ' . $response->message()
    );
  }

  return $response->content();
}

sub _url {
  my ( $self, $path, $params ) = @_;

  my $url = $self->{_cas}->clone();

  $url->path( $url->path() . $path );
  $url->query_param_append( $_ => $params->{$_} )
    for keys %$params;

  return $url;
}

sub _v20_validate {
  my ( $self, $path, $service, $ticket, %args ) = @_;

  my %params = (
    service => $service,
    ticket  => $ticket,
  );
  $params{renew} = 'true'
    if $args{renew};
  $params{pgtUrl} = URI->new( $args{pgtUrl} )->canonical()
    if defined $args{pgtUrl};

  my $content = $self->_server_request( $path, \%params );
  return $content
    if ref $content;

  my $response = eval{ $self->_parse_auth_response( $content ) };
  return $self->_error( 'Failed to parse server response' )
    if $@;

  return $response;
}


#======================================================================
# public methods
#

sub login_url {
  my ( $self, $service, %args ) = @_;

  my %params = (
    service => $service,
  );
  for ( qw/ renew gateway / ) {
    $params{$_} = 'true', last
      if $args{$_};
  }

  return $self->_url( '/login', \%params )->canonical();
}

sub logout_url {
  my ( $self, %args ) = @_;

  my %params;
  $params{url} = $args{url}
    if defined $args{url};

  return $self->_url( '/logout', \%params )->canonical();
}

sub validate {
  my ( $self, $service, $ticket, %args ) = @_;

  my %params = (
    service => $service,
    ticket  => $ticket,
  );
  $params{renew} = 'true'
    if $args{renew};

  my $content = $self->_server_request( '/validate', \%params );
  return $content
    if ref $content;

  return $self->_error('Server sent an invalid response')
    unless $content =~ /^(yes|no)\n(.*)\n$/;

  my ( $yn, $user ) = ( $1, $2 );
  return Authen::CAS::Client::Response::AuthFailure->new( code => 'V10_AUTH_FAILURE' ),
    unless $yn eq 'yes';

  return Authen::CAS::Client::Response::AuthSuccess->new( user => $user );
}

sub service_validate {
  my ( $self, $service, $ticket, %args ) = @_;
  return $self->_v20_validate( '/serviceValidate', $service, $ticket, %args );
}

sub proxy_validate {
  my ( $self, $service, $ticket, %args ) = @_;
  return $self->_v20_validate( '/proxyValidate', $service, $ticket, %args );
}

sub proxy {
  my ( $self, $pgt, $target ) = @_;

  my %params = (
    pgt           => $pgt,
    targetService => URI->new( $target ),
  );

  my $content = $self->_server_request( '/proxy', \%params );
  return $content
    if ref $content;

  my $response = eval { $self->_parse_proxy_response( $content ) };
  return $self->_error( 'Failed to parse server response' )
    if $@;

  return $response;
}


1;

__END__

=head1 NAME

Authen::CAS::Client - Provides an easy-to-use interface for authentication using JA-SIG's Central Authentication Service

=head1 SYNOPSIS

  use Authen::CAS::Client;

  my $cas = Authen::CAS::Client->new( 'https://example.com/cas' );


  # generate an HTTP redirect to the CAS login URL
  my $r = HTTP::Response->new( 302 );
  $r->header( Location => $cas->login_url() );


  # generate an HTTP redirect to the CAS logout URL
  my $r = HTTP::Response->new( 302 );
  $r->header( Location => $cas->logout_url() );


  # validate a service ticket (CAS v1.0)
  my $r = $cas->validate( $service, $ticket );
  if( $r->is_success() ) {
    print "User authenticated as: ", $r->user(), "\n";
  }

  # validate a service ticket (CAS v2.0)
  my $r = $cas->service_validate( $service, $ticket );
  if( $r->is_success() ) {
    print "User authenticated as: ", $r->user(), "\n";
  }


  # validate a service/proxy ticket (CAS v2.0)
  my $r = $cas->proxy_validate( $service, $ticket );
  if( $r->is_success() ) {
    print "User authenticated as: ", $r->user(), "\n";
    print "Proxied through:\n";
    print "  $_\n"
      for $r->proxies();
  }


  # validate a service ticket and request a proxy ticket (CAS v2.0)
  my $r = $cas->service_validate( $server, $ticket, pgtUrl => $url );
  if( $r->is_success() ) {
    print "User authenticated as: ", $r->user(), "\n";

    unless( defined $r->iou() ) {
      print "Service validation for proxying failed\n";
    }
    else {
      print "Proxy granting ticket IOU: ", $r->iou(), "\n";

      ...
      # map IOU to proxy granting ticket via request to pgtUrl
      ...

      $r = $cas->proxy( $pgt, $target_service );
      if( $r->is_success() ) {
        print "Proxy ticket issued: ", $r->proxy_ticket(), "\n";
      }
    }
  }

=head1 DESCRIPTION

The C<Authen::CAS::Client> module provides a simple interface for
authenticating users using JA-SIG's CAS protocol.  Both CAS v1.0
and v2.0 are supported.

=head1 METHODS

=over 2

=item new $URL [, %ARGS]

new() creates an instance of an C<Authen::CAS::Client> object.  C<$URL>
refers to the CAS server's base URL.  C<%ARGS> may contain the
following optional parameter:

=over 4

=item * fatal =E<gt> $BOOLEAN

If this argument is true, the CAS client will C<die()> when an error
occurs and C<$@> will contain the error message.  Otherwise an
C<Authen::CAS::Client::Response::Error> object will be returned.  See
L<Authen::CAS::Client::Response> for more detail on response objects.

=back

=item login_url $SERVICE [, %ARGS]

login_url() returns the CAS server's login URL which can be used to
redirect users to start the authentication process.  C<$SERVICE> is the
service identifier that will be used during validation requests.
C<%ARGS> may contain the following optional parameters:

=over 4

=item * renew =E<gt> $BOOLEAN

This causes the CAS server to force a user to re-authenticate even if
an SSO session is already present for that user.

=item * gateway =E<gt> $BOOLEAN

This causes the CAS server to only rely on SSO sessions for authentication.
If an SSO session is not available for the current user, validation
will result in a failure.

=back

=item logout_url [%ARGS]

logout_url() returns the CAS server's logout URL which can be used to
redirect users to end authenticated sessions.  C<%ARGS> may contain
the following optional parameter:

=over 4

=item * url =E<gt> $URL

If present, the CAS server will present the user with a link to the given
URL once the user has logged out.

=back

=item validate $SERVICE, $TICKET [, %ARGS]

validate() attempts to validate a service ticket using the CAS v1.0 protocol.
C<$SERVICE> is the service identifier that was passed to the CAS server
during the login process.  C<$TICKET> is the service ticket that was
received after a successful authentication attempt.  Returns an appropriate
L<Authen::CAS::Response> object.  C<%ARGS> may contain the following optional
parameter:

=over 4

=item * renew =E<gt> $BOOLEAN

This will cause the CAS server to respond with a failure if authentication
validation was done via a CAS SSO session.

=back

=item service_validate $SERVICE, $TICKET [, %ARGS]

service_validate() attempts to validate a service ticket using the CAS v2.0
protocol.  This is similar to C<validate()>, but allows for greater
flexibility when there is a need for proxying authentication to back-end
services.  The C<$SERVICE> and C<$TICKET> parameters are the same as above.
Returns an appropriate L<Authen::CAS::Response> object.  C<%ARGS> may
contain the following optional parameters:

=over 4

=item * renew =E<gt> $BOOLEAN

This will cause the CAS server to respond with a failure if authentication
validation was done via a CAS SSO session.

=item * pgtUrl =E<gt> $URL

This tells the CAS server that a proxy ticket needs to be issued for
proxying authentication to a back-end service.  C<$URL> corresponds to
a callback URL that the CAS server will use to verify the service's
identity.  Per the CAS specification, this URL must be HTTPS.  If this
verification fails, normal validation will occur, but a proxy granting
ticket IOU will not be issued.

Also note that this call will block until the CAS server completes its
service verification attempt.  The returned proxy granting ticket IOU
can then be used to retrieve the proxy granting ticket that was passed
as a parameter to the given URL.

=back

=item proxy_validate $SERVICE, $TICKET [, %ARGS]

proxy_validate() is almost identical in operation to C<service_validate()>
except that both service tickets and proxy tickets can be used for
validation and a list of proxies will be provided if proxied authentication
has been used.  The C<$SERVICE> and C<$TICKET> parameters are the same as
above.  Returns an appropriate L<Authen::CAS::Response> object.  C<%ARGS>
may contain the following optional parameters:

=over 4

=item * renew =E<gt> $BOOLEAN

This is the same as described above.

=item * pgtUrl =E<gt> $URL

This is the same as described above.

=back

=item proxy $PGT, $TARGET

proxy() is used to retrieve a proxy ticket that can be passed to a back-end
service for proxied authentication.  C<$PGT> is the proxy granting ticket
that was passed as a parameter to the C<pgtUrl> specified in either
C<service_validate()> or C<proxy_validate()>.  C<$TARGET> is the
service identifier for the back-end system that will be using the
returned proxy ticket for validation.  Returns an appropriate
L<Authen::CAS::Response> object.

=back

=head1 BUGS

None are known at this time, but if you find one, please feel free to
submit a report to the author.

=head1 AUTHOR

jason hord E<lt>pravus@cpan.orgE<gt>

=head1 SEE ALSO

L<Authen::CAS::Client::Response>

More information about CAS can be found at JA-SIG's CAS homepage:
L<http://www.ja-sig.org/products/cas/>

=head1 COPYRIGHT

Copyright (c) 2007-2009, jason hord

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:

=over 2

=item *

Redistributions of source code must retain the above copyright
notice, this list of conditions and the following disclaimer.

=item *

Redistributions in binary form must reproduce the above
copyright notice, this list of conditions and the following
disclaimer in the documentation and/or other materials provided
with the distribution.

=back

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut
